File: src/ZoneCodeGenerator/Domain/Evaluation/Operation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ZoneCodeGenerator.Domain.Evaluation
{
    class Operation : IEvaluation
    {
        public IEvaluation Operand1 { get; }
        public bool Operand1NeedsParenthesis => Operand1 is Operation operation1 &&
                                                operation1.OperationType.Precedence > OperationType.Precedence;

        public IEvaluation Operand2 { get; }
        public bool Operand2NeedsParenthesis => Operand2 is Operation operation2 &&
                                                operation2.OperationType.Precedence >= OperationType.Precedence;

        public OperationType OperationType { get; }

        public bool IsStatic => Operand1.IsStatic && Operand2.IsStatic;

        public int EvaluateNumeric()
        {
            return OperationType.Function(Operand1.EvaluateNumeric(), Operand2.EvaluateNumeric());
        }

        public Operation(IEvaluation operand1, IEvaluation operand2, OperationType type)
        {
            Operand1 = operand1;
            Operand2 = operand2;
            OperationType = type;
        }

        public override string ToString()
        {
            string operand1String;
            string operand2String;

            if (Operand1 is Operation operation1
                && operation1.OperationType.Precedence > OperationType.Precedence)
            {
                operand1String = $"({Operand1})";
            }
            else
            {
                operand1String = Operand1.ToString();
            }

            if (Operand2 is Operation operation2
                && operation2.OperationType.Precedence > OperationType.Precedence)
            {
                operand2String = $"({Operand2})";
            }
            else
            {
                operand2String = Operand2.ToString();
            }

            return $"{operand1String} {OperationType} {operand2String}";
        }
    }
}

File: src/ZoneCodeGeneratorLib/Domain/Evaluation/Operation.cpp
#include "Operation.h"

Operation::Operation(const OperationType* type)
    : m_operation_type(type)
{
}

Operation::Operation(const OperationType* type, std::unique_ptr<IEvaluation> operand1, std::unique_ptr<IEvaluation> operand2)
    : m_operation_type(type),
      m_operand1(std::move(operand1)),
      m_operand2(std::move(operand2))
{
}

EvaluationType Operation::GetType() const
{
    return EvaluationType::OPERATION;
}

bool Operation::IsStatic() const
{
    return m_operand1->IsStatic() && m_operand2->IsStatic();
}

int Operation::EvaluateNumeric() const
{
    return m_operation_type->m_evaluation_function(m_operand1->EvaluateNumeric(), m_operand2->EvaluateNumeric());
}

bool Operation::Operand1NeedsParenthesis() const
{
    return m_operand1->GetType() == EvaluationType::OPERATION
        && dynamic_cast<Operation*>(m_operand1.get())->m_operation_type->m_precedence > m_operation_type->m_precedence;
}

bool Operation::Operand2NeedsParenthesis() const
{
    return m_operand2->GetType() == EvaluationType::OPERATION
        && dynamic_cast<Operation*>(m_operand2.get())->m_operation_type->m_precedence > m_operation_type->m_precedence;
}
